### ISSUE TITLE
MQTT notifications

### ISSUE THREAD

**Issue by fgalan** [telefonicaid/fiware-orion#3001]
This feature is about developing MQTT notifications in CB, in addition to the extisting ones in HTTP. The work will be based in the one done at branch https://github.com/telefonicaid/fiware-orion/tree/feature/mqtt_notifications_poc which already implemented a MQTT notification system (but very limited, as it is hardwired).

The payload used by MQTT notification will be the same one used for NGSIv2 notifications, i.e.: normalized, keyValues, values, legacy or customiced.

From an API point of view is is a matter of using `mqtt/mqttCustom` instead of existing `http/httpCustom` fields in subscriptions. E.g.:

```
"mqtt": {
  "topic": "/my/topic"
}
```

or 

```
"mqttCustom": {
  "topic": "/entity/{$id}/topic",
  "payload": "The temperature value is ${temperature}"
}
```

From a CLI point of view, a new parameter will be used to set MQTT broker host+port, e.g: `-mqttbroker` (similar to existing `-dbhost`). This parameter also acts as a switch to enable/disable MQTT notifications: if `-mqttbroker` is not defined then MQTT notifications are ignored.

Note that MQTT protocol doesn't have any mechanism similar to HTTP headers. Thus, Fiware-Service, Fiware-ServicePath, X-Auth-Token and Fiware-Correlator cannot be directly included in MQTT notifications. There are three approachs to cope with this:

1. Ignore them, assuming that the kind of receivers for MQTT notifications are not interested in that kind of inoformation.
2. Include that information in the MQTT payload as "virtual attributes" (in the same way dateCreated, dataModified, etc. works)
3. Modify the notification templating mechanisms defining ${fiware-service}, ${fiware-servicepath}, etc. tokens, so the client can use in the places of the notification (topic or payload fields) where to include.

**Comment by McMutton** [comment]
Hi @fgalan, I have just started working on this issue. In case I have any questions, I'll post here, thank you.

**Comment by McMutton** [comment]
Hi @fgalan, I have gathered enough information from the issue description, existing documents and the code in the task branch to start the implementation. 

Before doing so, I would like to share my work plan with you to check if that is OK with you. Instead of sending a huge pull request at the end, I would like to divide the task into several ones as below.

First PR: 
- MQTT CLI Switch
- Minor changes/improvements on the core MQTT handler
- Tests

Second PR:
- mqttCustom notifications
- Tests

Third PR:
- Handling of HTTP Headers 
- Tests

Please let me know if you prefer to have it in smaller number of PRs or if you have any other comments.

**Comment by fgalan** [comment]
Thanks for the feedback!

The idea to have incremantal PRs makes sense to me in this case. Thus, instead of using master as base branch for these PRs, let's use a feature/XXXX branch. That feature/XXXX could be the existing feature/mqtt_notifications_poc  (which has been update with master recently) or a fresh new one if you prefer to start from scratch (I prefer use feature/mqtt_notifications_poc but if you think different, tell me so and I'd create a new feature/XXXX branch for you).

The procedure is to start each piece of work as a task/YYYY branch branching from feature/XXXX, then cast PR for task/YYYY using feature/XXXX as base branch. Once the feature is complete, feature/XXXX is merged into master with a final PR (which doesn't need to much review, as all the bits has been already reviewed in the incremental PRs for task/YYYY branches).

With regards the plan, the context of each one of the 3 steps you mention makes sense. Don't need of additional PRs, but let's see as we go. With regard to third PR we need to decide which option to chosee (between 2 or 3 above). However, let's wait until second PR is on the way to make that decission ;)

**Comment by McMutton** [comment]
Agreed, no need to create a new base branch, I'll use _feature/mqtt_notifications_poc_.

> With regard to third PR we need to decide which option to chosee (between 2 or 3 above). However, let's wait until second PR is on the way to make that decission ;) 

Sure, we can discuss it in detail before starting the third PR.

**Comment by fgalan** [comment]
Step 1 done in PR https://github.com/telefonicaid/fiware-orion/pull/3194

**Comment by cdupont** [comment]
This is highly needed.
I need to implement MQTT on my server, which is using Orion as a back-end.
This proves difficult if MQTT is not implemented inside Orion... For instance I could implement myself a PUBLISH endpoint and write inside the MongoDB directly. But in this case the Orion notifications would not be triggered I think.

**Comment by fgalan** [comment]
I agree this is a very apealing to be included in Orion.

The work was started in PR #3194. As far as I remember the work was pretty advanced, but some comments on the PR are yet pending. Not sure if @McMutton are you still active and can provide some update on it, please.

If @McMutton is no longer involved in Orion development, a possibility will be that some other continues the work he started with a "parallel" PR including all his code + pending parts. Do you want to volunteer @cdupont ? ;)

**Comment by cdupont** [comment]
Unfortunately my time is too limited currently :)

**Comment by fgalan** [comment]
A long time has passed since last activity in this issue :)

I have to assume that @McMutton is missing in action :( so I'll take this issue myself.

After reviewing the original specification and some feedback provided by @kzangeli [here](https://github.com/telefonicaid/fiware-orion/pull/3194) what to be implemente will be the following:

* `mqtt` and `mqttCustom` in parallel to current `http` and `httpCustom`
* as MQTT basic information (existing in both `mqtt` and `mqttCustom` the following fields will be used:
    * `endpoint`, for the MQTT broker endpoint in the `host:port` format
    * `topic`, for the MQTT topic. It is better not include this as part of the endpoint so a future alias system for endpoints could fit better
    * `qos`
* The `mqttCustom` object adds also a `payload` field to precisely define the payload to be sent (the same used in `httpCustom`). If this is not used, standard normalized NGSIv2 format will be sent.
* The `mqttCustom` will support the macro system `${...}` (the same used in `httpCustom`)
     * Does this macro system supports $service and $subservice or similar? (To be checked)

Connections to MQTT brokers will be reused among subscriptions.

**Comment by fgalan** [comment]
I'll have a look to existing PR https://github.com/telefonicaid/fiware-orion/pull/3194 and also the telefonicaid:feature/mqtt_notifications_poc branch and probably reuse some pieces of work there.

**Comment by fgalan** [comment]
> * Does this macro system supports $service and $subservice or similar? (To be checked)

From [NGSIv2 spec](http://telefonicaid.github.io/fiware-orion/api/v2/stable/) "Custom notifications" section:

> Macro substitution for templates is based on the syntax ${..}. In particular:
>
> * `${id}` is replaced by the id of the entity
> * `${type}` is replaced by the type of the entity
> Any other `${token}` is replaced by the value of the attribute whose name is token or with an empty string if the attribute is not included in the notification. If the value is a number, a bool or null then its string representation is used. If the value is a JSON array or object then its JSON representation as string is used.

Looking https://fiware-orion.readthedocs.io/en/master/user/ngsiv2_implementation_notes/index.html is not adding any new `${...}` to that.

Thus, it seems that macro for service and subservice are not yet implemented.

**Comment by fgalan** [comment]
Done in PR https://github.com/telefonicaid/fiware-orion/pull/3888

Follow up issues with MQTT improvements:

* #3913 
* #3914 
* #3915
